### PR TITLE
[Identity-based Authorization] Add an option to specify a policy for …

### PR DESCRIPTION
**Pull request by peusebiu**
…all repositories

### CODE REVIEW

**Review by shimish2 — COMMENTED** [review]
How we are dealing with this particular scenario:- 

Suppose for user **charlie** we specified an all repository policy and allowed action is **create** only and for another repository e.g repo1 a policy contains **charlie** user and permitted action is ["create","update","delete","read"].

 Are we allowing other actions apart from create  for user **charlie**  for repo1 ? 

e.g config
``` json 
"accessControl": {
    "/": {
      "policies": [
        {
          "users": ["charlie"],
          "actions": ["create"]
        }
      ],
      "defaultPolicy": ["read"]
    },
    "repos1/repo": {
        "policies": [
        {
            "users": ["alice", "bob","charlie"],
            "actions": ["create", "read", "update", "delete"]
        },
        {
            "users": ["mallory"],
            "actions": ["create", "read"]
        }
        ],
        "defaultPolicy": ["read"]
    }
  ```

**Comment by shimish2** [comment]
> Yes, that's what I thought about, in the current implementation we indeed allow charlie to write/read/update/delete on repo1.
> Basically we just make a union with all the policies.
> 
> 1. If user has perm "X" on a repo from all-repositories based policy then allow otherwise 2
> 2. if not 1 then check repository based policy then allow otherwise 3
> 3. if not 2 then check admin based policy then allow otherwise don't allow
> 
> Is this ok?

@rchincha should we do intersection of all policies or current implementation is fine.

**Comment by peusebiu** [comment]
> Another comment is that /v2/ must always be available to folks without authn or authz because rest of the bootstrapping will happen based on that. Is this true? Is there a test case for it?

Yeah, it's covered for authz but not for authn. I had a testcase but I was assuming that clients would give username:password on /v2/ route.
I will update to also cover authn logic.

File: pkg/api/controller_test.go [hunk 1]
```diff
@@ -1666,6 +1666,15 @@ func TestAuthorizationWithBasicAuth(t *testing.T) {
 					},
 					DefaultPolicy: []string{},
 				},
+				"globalpolicy": api.PolicyGroup{
```
Comment (rchincha) [inline pkg/api/controller_test.go:1669]: typo "globalpolicy"

File: pkg/api/authz.go [hunk 2]
```diff
@@ -79,12 +79,20 @@ func (ac *AccessController) getReadRepos(username string) []string {
 // can verifies if a user can do action on repository.
 func (ac *AccessController) can(username, action, repository string) bool {
 	can := false
-	// check repo based policy
-	pg, ok := ac.Config.Repositories[repository]
+
+	pg, ok := ac.Config.Repositories["globalpolicy"]
```
Comment (rchincha) [inline pkg/api/authz.go:83]: Make "globalpolicy" a string constant

**Review by rchincha — COMMENTED** [review]
let me evaluate whether we should just have "/" section or this "globalpolicy". Also whether these paths should be regexps, and we have a "*' or a "/*"

File: pkg/api/controller_test.go (hunk 1 above)
Comment (peusebiu) [reply pkg/api/controller_test.go]: it's not, viper is not case sensitive.

**Comment by peusebiu** [comment]
@rchincha  rebased, ready to be reviewed.

File: examples/README.md [hunk 3]
```diff
@@ -186,9 +186,24 @@ identities. An additional per-repository default policy can be specified for
...
+Regexes can also be used as repository paths.
+
+Authorization is granted based on the longest path matched.
+For example repos2/repo repository will match both ".*" and "repos2/repo" keys,
+in such case repos2/repo policy will be used.
+
 ```
 "accessControl": {
-    "repos1/repo": {
+    ".*": {
```
Comment (rchincha) [inline examples/README.md:197]: Can we use the same pattern we use for "sync" feature?
\* and ** for subdirs?

**Review by rchincha — COMMENTED** [review]
Here are the test cases we need to cover:
1) developer test repos - any user can create a repo under a repo prefix
2) production repos - admin users need to first create a repo and then some users can write to this repo, but cannot delete
3) admin users - some admin users can do anything

File: examples/README.md (hunk 3 above)
Comment (peusebiu) [reply examples/README.md]: Added glob patterns

**Review by rchincha — APPROVED** [review]
lgtm

**Pull request merged**